transmart-api-server/src/main/groovy/org/transmartproject/api/server/client/CustomClientRequestFactory.groovy

```diff
@@ -0,0 +1,64 @@
+package org.transmartproject.api.server.client
+
+import org.apache.http.client.methods.HttpUriRequest
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.beans.factory.config.ConfigurableBeanFactory
+import org.springframework.context.annotation.Scope
+import org.springframework.http.*
+import org.springframework.http.client.ClientHttpRequestFactory
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory
+import org.springframework.stereotype.Component
+import org.springframework.util.LinkedMultiValueMap
+import org.springframework.util.MultiValueMap
+import org.springframework.web.client.RestTemplate
+
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+class CustomClientRequestFactory extends HttpComponentsClientHttpRequestFactory implements ClientHttpRequestFactory {
+
+    @Value('${keycloak.clientId}')
+    private String clientId
+
+    @Value('${keycloak.offlineToken}')
+    private String offlineToken
+
+    @Value('${keycloak.realm}')
+    private String realm
+
+    @Value('${keycloak.serverUrl}')
+    private String keycloakServerUrl
+
+    public static final String AUTHORIZATION_HEADER = "Authorization"
+
+
+    @Override
+    protected void postProcessHttpRequest(HttpUriRequest request) {
+        String offlineToken = getOfflineAccessToken()
+        request.setHeader(AUTHORIZATION_HEADER, "Bearer " + offlineToken)
+    }
+
+    /**
+     * Get access token from Keycloak based on the client_id and the offline token, which is stored in the config.
+     * Offline token is a type of a classic Refresh token, but it never expires.
+     * @return access token
+     */
+    private String getOfflineAccessToken(){
```

Comment:
AFAIK offline token gives us right to get a new access token. `getNewAccessTokenByOfflineTokenAndClientId()` express the purpose of the method better, more mouthful though.